File: src/serialization.ts
/**
 * Serialization helpers for bridging JSON ↔ typed objects
 *
 * Data crosses the native bridge as JSON strings. These helpers
 * convert between our TypeScript types and the JSON payloads
 * expected by the native modules.
 */

import type {
  KoraIDVConfiguration,
  Verification,
  VerificationFlowResult,
  DocumentVerification,
  FaceVerification,
  LivenessVerification,
  ChallengeResult,
  RiskSignal,
} from './types';
import { KoraError, KoraErrorCode } from './types';

// ---------------------------------------------------------------------------
// Configuration → JSON (sent to native)
// ---------------------------------------------------------------------------

export function serializeConfiguration(config: KoraIDVConfiguration): string {
  const payload: Record<string, unknown> = {
    apiKey: config.apiKey,
    tenantId: config.tenantId,
  };

  if (config.environment !== undefined) {
    payload.environment = config.environment;
  }
  if (config.baseUrl !== undefined) {
    payload.baseUrl = config.baseUrl;
  }
  if (config.documentTypes !== undefined) {
    payload.documentTypes = config.documentTypes;
  }
  if (config.livenessMode !== undefined) {
    payload.livenessMode = config.livenessMode;
  }
  if (config.theme !== undefined) {
    payload.theme = config.theme;
  }
  if (config.timeout !== undefined) {
    payload.timeout = config.timeout;
  }
  if (config.debugLogging !== undefined) {
    payload.debugLogging = config.debugLogging;
  }

  return JSON.stringify(payload);
}

// ---------------------------------------------------------------------------
// JSON → Verification result (received from native)
// ---------------------------------------------------------------------------

export function deserializeResult(json: string): VerificationFlowResult {
  const raw = JSON.parse(json) as Record<string, unknown>;

  if (raw.type === 'cancelled') {
    return { type: 'cancelled' };
  }

  if (raw.type === 'success') {
    const verification = deserializeVerification(
      raw.verification as Record<string, unknown>
    );
    return { type: 'success', verification };
  }

  throw new KoraError(
    KoraErrorCode.INVALID_RESPONSE,
    `Unknown result type: ${String(raw.type)}`
  );
}

// ---------------------------------------------------------------------------
// JSON → KoraError (received from native Promise rejection)
// ---------------------------------------------------------------------------

export function deserializeError(code: string, message: string): KoraError {
  const errorCode = isKoraErrorCode(code) ? code : KoraErrorCode.UNKNOWN;
  return new KoraError(errorCode, message);
}

function isKoraErrorCode(code: string): code is KoraErrorCode {
  return Object.values(KoraErrorCode).includes(code as KoraErrorCode);
}

// ---------------------------------------------------------------------------
// Verification deserialization
// ---------------------------------------------------------------------------

function deserializeVerification(raw: Record<string, unknown>): Verification {
  return {
    id: raw.id as string,
    externalId: raw.externalId as string,
    tenantId: raw.tenantId as string,
    tier: raw.tier as string,
    status: raw.status as Verification['status'],
    documentVerification: raw.documentVerification
      ? deserializeDocumentVerification(
          raw.documentVerification as Record<string, unknown>
        )
      : null,
    faceVerification: raw.faceVerification
      ? deserializeFaceVerification(
          raw.faceVerification as Record<string, unknown>
        )
      : null,
    livenessVerification: raw.livenessVerification
      ? deserializeLivenessVerification(
          raw.livenessVerification as Record<string, unknown>
        )
      : null,
    riskSignals: raw.riskSignals
      ? (raw.riskSignals as Record<string, unknown>[]).map(
          deserializeRiskSignal
        )
      : null,
    riskScore: (raw.riskScore as number) ?? null,
    createdAt: raw.createdAt as string,
    updatedAt: raw.updatedAt as string,
    completedAt: (raw.completedAt as string) ?? null,
  };
}

function deserializeDocumentVerification(
  raw: Record<string, unknown>
): DocumentVerification {
  return {
    documentType: raw.documentType as string,
    documentNumber: (raw.documentNumber as string) ?? null,
    firstName: (raw.firstName as string) ?? null,
    lastName: (raw.lastName as string) ?? null,
    dateOfBirth: (raw.dateOfBirth as string) ?? null,
    expirationDate: (raw.expirationDate as string) ?? null,
    issuingCountry: (raw.issuingCountry as string) ?? null,
    mrzValid: (raw.mrzValid as boolean) ?? null,
    authenticityScore: (raw.authenticityScore as number) ?? null,
    extractedFields:
      (raw.extractedFields as Record<string, string>) ?? null,
  };
}

function deserializeFaceVerification(
  raw: Record<string, unknown>
): FaceVerification {
  return {
    matchScore: raw.matchScore as number,
    matchResult: raw.matchResult as string,
    confidence: raw.confidence as number,
  };
}

function deserializeLivenessVerification(
  raw: Record<string, unknown>
): LivenessVerification {
  return {
    livenessScore: raw.livenessScore as number,
    isLive: raw.isLive as boolean,
    challengeResults: raw.challengeResults
      ? (raw.challengeResults as Record<string, unknown>[]).map(
          deserializeChallengeResult
        )
      : null,
  };
}

function deserializeChallengeResult(
  raw: Record<string, unknown>
): ChallengeResult {
  return {
    type: raw.type as string,
    passed: raw.passed as boolean,
    confidence: raw.confidence as number,
  };
}

function deserializeRiskSignal(raw: Record<string, unknown>): RiskSignal {
  return {
    code: raw.code as string,
    severity: raw.severity as string,
    message: raw.message as string,
  };
}

File: android/src/main/kotlin/com/koraidv/reactnative/Serialization.kt
package com.koraidv.reactnative

import com.koraidv.sdk.ChallengeResult
import com.koraidv.sdk.DocumentVerification
import com.koraidv.sdk.FaceVerification
import com.koraidv.sdk.LivenessVerification
import com.koraidv.sdk.RiskSignal
import com.koraidv.sdk.Verification
import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * JSON serialization helpers for crossing the React Native bridge.
 *
 * Converts native SDK models → JSON strings that the TypeScript
 * deserialization layer can parse.
 */
object Serialization {

    private val iso8601Format: SimpleDateFormat
        get() = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }

    /**
     * Build a success result JSON string containing the verification.
     */
    fun buildSuccessResult(verification: Verification): String {
        val result = JSONObject().apply {
            put("type", "success")
            put("verification", serializeVerification(verification))
        }
        return result.toString()
    }

    /**
     * Serialize a Verification to JSONObject.
     */
    fun serializeVerification(v: Verification): JSONObject {
        return JSONObject().apply {
            put("id", v.id)
            put("externalId", v.externalId)
            put("tenantId", v.tenantId)
            put("tier", v.tier)
            put("status", v.status.value)

            if (v.documentVerification != null) {
                put("documentVerification", serializeDocumentVerification(v.documentVerification))
            }
            if (v.faceVerification != null) {
                put("faceVerification", serializeFaceVerification(v.faceVerification))
            }
            if (v.livenessVerification != null) {
                put("livenessVerification", serializeLivenessVerification(v.livenessVerification))
            }
            if (v.riskSignals != null) {
                put("riskSignals", serializeRiskSignals(v.riskSignals))
            }
            putOpt("riskScore", v.riskScore)
            put("createdAt", formatDate(v.createdAt))
            put("updatedAt", formatDate(v.updatedAt))
            putOpt("completedAt", v.completedAt?.let { formatDate(it) })
        }
    }

    private fun serializeDocumentVerification(dv: DocumentVerification): JSONObject {
        return JSONObject().apply {
            put("documentType", dv.documentType)
            putOpt("documentNumber", dv.documentNumber)
            putOpt("firstName", dv.firstName)
            putOpt("lastName", dv.lastName)
            putOpt("dateOfBirth", dv.dateOfBirth)
            putOpt("expirationDate", dv.expirationDate)
            putOpt("issuingCountry", dv.issuingCountry)
            putOpt("mrzValid", dv.mrzValid)
            putOpt("authenticityScore", dv.authenticityScore)
            if (dv.extractedFields != null) {
                put("extractedFields", JSONObject(dv.extractedFields))
            }
        }
    }

    private fun serializeFaceVerification(fv: FaceVerification): JSONObject {
        return JSONObject().apply {
            put("matchScore", fv.matchScore)
            put("matchResult", fv.matchResult)
            put("confidence", fv.confidence)
        }
    }

    private fun serializeLivenessVerification(lv: LivenessVerification): JSONObject {
        return JSONObject().apply {
            put("livenessScore", lv.livenessScore)
            put("isLive", lv.isLive)
            if (lv.challengeResults != null) {
                val arr = JSONArray()
                for (cr in lv.challengeResults) {
                    arr.put(serializeChallengeResult(cr))
                }
                put("challengeResults", arr)
            }
        }
    }

    private fun serializeChallengeResult(cr: ChallengeResult): JSONObject {
        return JSONObject().apply {
            put("type", cr.type)
            put("passed", cr.passed)
            put("confidence", cr.confidence)
        }
    }

    private fun serializeRiskSignals(signals: List<RiskSignal>): JSONArray {
        val arr = JSONArray()
        for (signal in signals) {
            arr.put(JSONObject().apply {
                put("code", signal.code)
                put("severity", signal.severity)
                put("message", signal.message)
            })
        }
        return arr
    }

    private fun formatDate(date: Date): String {
        return iso8601Format.format(date)
    }
}
